src/main/java/com/medallia/word2vec/neuralnetwork/NeuralNetworkTrainer.java

```diff
@@ -151,28 +156,54 @@ static long incrementRandom(long r) {
 	}
 	
 	/** @return Trained NN model */
-	public NeuralNetworkModel train(Iterable<List<String>> sentences) throws InterruptedException {
-		ListeningExecutorService ex = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(config.numThreads));
-		
+	public NeuralNetworkModel train(final Iterable<List<String>> sentences) throws InterruptedException {
+		final ListeningExecutorService ex =
+				MoreExecutors.listeningDecorator(
+					new ThreadPoolExecutor(config.numThreads, config.numThreads,
+							0L, TimeUnit.MILLISECONDS,
+							new ArrayBlockingQueue<Runnable>(8),
+							new ThreadPoolExecutor.CallerRunsPolicy()));
+
 		int numSentences = Iterables.size(sentences);
 		numTrainedTokens += numSentences;
-		
-		// Partition the sentences evenly amongst the threads
-		Iterable<List<List<String>>> partitioned = Iterables.partition(sentences, numSentences / config.numThreads + 1);
-		
+
+		// Partition the sentences into batches
+		final Iterable<List<List<String>>> batched = new Iterable<List<List<String>>>() {
+			@Override public Iterator<List<List<String>>> iterator() {
+				return new Iterator<List<List<String>>>() {
+					private final Iterator<List<String>> inner = sentences.iterator();
+
+					@Override
+					public boolean hasNext() {
+						return inner.hasNext();
+					}
+
+					@Override
+					public List<List<String>> next() {
+						if(!hasNext())
+							throw new NoSuchElementException();
+
+						return Lists.newArrayList(Iterators.limit(inner, 1024));
+					}
+
+					@Override
+					public void remove() {
+						throw new UnsupportedOperationException();
+					}
+				};
+			}
+		};
+
 		try {
 			listener.update(Stage.TRAIN_NEURAL_NETWORK, 0.0);
 			for (int iter = config.iterations; iter > 0; iter--) {
-				List<CallableVoid> tasks = new ArrayList<>();
+				List<ListenableFuture<?>> futures = new ArrayList<>(64);
 				int i = 0;
-				for (final List<List<String>> batch : partitioned) {
-					tasks.add(createWorker(i, iter, batch));
+				for (final List<List<String>> batch : batched) {
+					futures.add(ex.submit(createWorker(i, iter, batch)));
 					i++;
 				}
```

Comment:
This `for` loop would pull the entire training data into memory, because every worker contains a batch, and all workers are instantiated before the first one starts working.